### ISSUE TITLE
There seems to be no supported way to host a tracker and a client in a single process and let them communicate

### ISSUE THREAD

**Issue by ManlyMarco** [alanmcgovern/monotorrent#600]
My app generates a private torrent file and hosts a tracker for it, as well as a client that seeds the torrent.

The issue is that there doesn't seem to be any easy way to tell the tracker about this (the client opens the torrent and sees its own external IP address as a tracker, so when it tries to connect to the tracker the connection never reaches).

The only way I've found to make this work is use TrackerManager.AddTrackerAsync to add localhost tracker to the torrent, and then set ReportedAddress to the external IP so that peers see the right address.

The issue with this solution is that I want to have a private torrent, which stops AddTrackerAsync from working. (I did get it to work by patching out the private check but that's a hack not a real solution)

Is there any better way of doing this?

**Comment by alanmcgovern** [comment]
Could you try something like this to rewrite the `Url` before the engine creates the tracker object?

```
// If you're customising the Factories object already, just customise these two extra things
// Otherwise you should start with the default Factories object. I'm assuming you're not
// customising things already...
var defaultFactories = Factories.Default;

// Set up a redirect to replace 'mytracker.com' with '127.0.0.1' for http trackers.
var factories = defaultFactories.WithTrackerCreator ("http", uri => {
    if (uri.Host == "mytracker.com")
        uri = new UriBuilder (uri) { Host = "127.0.0.1" }.Uri;
    return defaultFactories.CreateTracker (uri);
});

// Add a redirect to replace 'mytracker.com' with '127.0.0.1' for http trackers.
factories = factories.WithTrackerCreator ("https", uri => {
    if (uri.Host == "mytracker.com")
        uri = new UriBuilder (uri) { Host = "127.0.0.1" }.Uri;
    return defaultFactories.CreateTracker (uri);
});

var engine = new ClientEngine (new EngineSettings (), factories);

// Do your regular stuff...
```

If that doesn't work, let me know!

**Comment by ManlyMarco** [comment]
Testing got blocked by #610

**Comment by ManlyMarco** [comment]
Can confirm that this is working fine https://github.com/alanmcgovern/monotorrent/issues/600#issuecomment-1404364900
I'm not sure how often people would need to do this, but it'd be nice to have a simpler and more obvious way of doing this.

**Comment by ManlyMarco** [comment]
I'm not sure if it's a side effect of this or something that always happens, but the seed keeps trying to connect to itself every couple of minutes, failing with HandshakeFailed every time.
